src/Search/Dynamic/Algorithm/Improved.php

```diff
@@ -0,0 +1,92 @@
+<?php
+/**
+ * ScandiPWA_CatalogGraphQl
+ *
+ * @category ScandiPWA
+ * @package ScandiPWA_CatalogGraphQl
+ * @author Daniels Puzina <info@scandiweb.com>
+ * @copyright Copyright (c) 2020 Scandiweb, Ltd (https://scandiweb.com)
+ */
+namespace ScandiPWA\CatalogGraphQl\Search\Dynamic\Algorithm;
+
+use Magento\Framework\DB\Ddl\Table;
+use Magento\Framework\Search\Adapter\OptionsInterface;
+use Magento\Framework\Search\Dynamic\Algorithm;
+use Magento\Framework\Search\Dynamic\Algorithm\AlgorithmInterface;
+use Magento\Framework\Search\Dynamic\DataProviderInterface;
+use Magento\Framework\Search\Request\BucketInterface;
+
+class Improved implements AlgorithmInterface
+{
+    /**
+     * @var Algorithm
+     */
+    protected $algorithm;
+
+    /**
+     * @var DataProviderInterface
+     */
+    protected $dataProvider;
+
+    /**
+     * @var OptionsInterface
+     */
+    protected $options;
+
+    /**
+     * @param DataProviderInterface $dataProvider
+     * @param Algorithm $algorithm
+     * @param OptionsInterface $options
+     */
+    public function __construct(
+        DataProviderInterface $dataProvider,
+        Algorithm $algorithm,
+        OptionsInterface $options
+    ) {
+        $this->algorithm = $algorithm;
+        $this->dataProvider = $dataProvider;
+        $this->options = $options;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getItems(
+        BucketInterface $bucket,
+        array $dimensions,
+        \Magento\Framework\Search\Dynamic\EntityStorage $entityStorage
+    ) {
+        $aggregations = $this->dataProvider->getAggregations($entityStorage);
+
+        $options = $this->options->get();
+        if ($aggregations['count'] < $options['interval_division_limit']) {
+            return [[
+                'from' => $aggregations['min'],
+                'to' => $aggregations['max'],
+                'count' => $aggregations['count']
+            ]];
+        }
+        $this->algorithm->setStatistics(
+            $aggregations['min'],
+            $aggregations['max'],
+            $aggregations['std'],
+            $aggregations['count']
+        );
+
+        $this->algorithm->setLimits($aggregations['min'], $aggregations['max']);
+
+        $interval = $this->dataProvider->getInterval($bucket, $dimensions, $entityStorage);
+        $data = $this->algorithm->calculateSeparators($interval);
+
+        $data[0]['from'] = 0;
+
+        $dataSize = count($data);
+        for ($key = 0; $key < $dataSize; $key++) {
+            if (isset($data[$key + 1])) {
+                $data[$key]['to'] = $data[$key + 1]['from'];
+            }
+        }
```

Comment:
Better to use foreach here

Reply:
I'd suggest leaving this piece of code as is, because it was copied completely from magento codebase unchanged. If we don't refactor copy-pasted code, it will be simpler for other developers to compare methods  and understand was changed to fix issues.

Reply:
In this case you don't change code, you make it simpler to read and foreach is better in performance vs for. So I considering to change to be more readable even if it is magento base code.